### ISSUE TITLE
Default constructor created for non-default constructible type

### ISSUE THREAD

**Issue by praetorian20** [CastXML/CastXML#55]
Given this type
```
struct const_item
{
  const int values[10];
};
```
CastXML generates the following XML output
```
  <Struct id="_6" name="const_item" context="_1" location="f1:578" file="f1" line="578" members="_11 _12 _13 _14" size="320" align="32"/>
  <Constructor id="_12" name="const_item" context="_6" access="public" location="f1:578" file="f1" line="578" inline="1" artificial="1"/>
  <Constructor id="_13" name="const_item" context="_6" access="public" location="f1:578" file="f1" line="578" inline="1" artificial="1" throw="">
    <Argument type="_20" location="f1:578" file="f1" line="578"/>
  </Constructor>
```
The first `Constructor` element seems invalid because `const_item`'s default constructor is implicitly declared as deleted. Otherwise, the implicit definition would be ill-formed because it would attempt to default initialize the `const` data member. This is specified in the last sentence under [[dcl.init]/7](http://eel.is/c++draft/dcl.init#7)

> If a program calls for the default initialization of an object of a const-qualified type `T`, `T` shall be a class type with a user-provided default constructor.

Both gcc and clang fail to compile [this example](http://coliru.stacked-crooked.com/a/b60ca8d9032f4644).

The first `Constructor` element should either be absent, or it should have an extra attribute, say `deleted="1"`. If the extra attribute is added, it might be a good idea to also include another attribute `defaulted="1"` which would be set for C++11 code that explicitly `=default`s member functions.

**Comment by bradking** [comment]
CastXML has fairly elaborate infrastructure (and test cases) specifically to try to detect which implicitly members are valid.  One can see the code [here](https://github.com/CastXML/CastXML/blob/2edc188fd48a6a8467610d9d559985c4893c879a/src/RunClang.cxx#L71-L113) and [here](https://github.com/CastXML/CastXML/blob/2edc188fd48a6a8467610d9d559985c4893c879a/src/RunClang.cxx#L122-L129), and test cases [here](https://github.com/CastXML/CastXML/blob/2edc188fd48a6a8467610d9d559985c4893c879a/test/CMakeLists.txt#L285-L291).  The approach is to mark the implicit members as referenced (as if they were called) and suppress errors that Clang reports from doing so.  The resulting structure we get from Clang normally marks failed declarations as invalid so we skip generating them.  For some reason Clang is not doing that in this case.

Actually there is a test case close to the reported example [here](https://github.com/CastXML/CastXML/blob/2edc188fd48a6a8467610d9d559985c4893c879a/test/input/Class-implicit-member-const.cxx).  However, it has an explicit default constructor because otherwise Clang warns:

    .../test/input/Class-implicit-member-const.cxx:1:7: warning: class 'start' does not declare any constructor to initialize its non-modifiable members
    class start {
          ^
    .../test/input/Class-implicit-member-const.cxx:2:13: note: const member 'data' will never be initialized
      int const data;
                ^

The assumption made when the above-linked code was developed and test cases added was that this case is not useful in practice because it is not possible to use a class of this form.  With a `const` data member one must define at least one constructor explicitly in order to initialize the member.  Once such an explicit constructor is added then there is no implicit default constructor anyway.

What is the use case in practice for the example shown?

**Comment by praetorian20** [comment]
Brad, the `class start` example you pointed to is different from mine because my example has an [aggregate](http://eel.is/c++draft/dcl.init.aggr#1). `start` in your example is not an aggregate because it has a private data member, so clang is right to complain about the absence of a user defined constructor in that case.

An aggregate with a `const` data member is still useful because I can either value initialize or aggregate initialize an instance - [example](http://coliru.stacked-crooked.com/a/b8d2bed0f44136c8).

The problem with the current behavior is that it's causing some [tests in pygccxml](https://github.com/gccxml/pygccxml/pull/35#discussion_r59106513) to detect types such as the one in my example as *trivially default constructible* when they're actually not because of the presence of the `const` member.

**Comment by bradking** [comment]
Thanks, I hadn't noticed you're using an aggregate type.  It looks like Clang correctly deletes the default constructor for C++11 and above.  With `-std=c++11` and `-std=c++14` CastXML produces the correct output.  It is only for `-std=c++98` that the default constructor is incorrectly produced.

One can see that Clang treats the case differently depending on the language level:

```console
$ cat test.cxx                                                                                                                               
struct foo {
  int const bar;
};
void f() {
  foo f;
}
$ clang++ test.cxx -std=c++98
test.cxx:1:8: error: implicit default constructor for 'foo' must explicitly initialize the const member 'bar'
struct foo {
       ^
test.cxx:2:13: note: declared here
  int const bar;
            ^
test.cxx:5:7: note: implicit default constructor for 'foo' first required here
  foo f;
      ^
1 error generated.
$ clang++ test.cxx -std=c++11
test.cxx:5:7: error: call to implicitly-deleted default constructor of 'foo'
  foo f;
      ^
test.cxx:2:13: note: default constructor of 'foo' is implicitly deleted because field 'bar' of const-qualified type 'const int' would not be initialized
  int const bar;
            ^
1 error generated.
```

**Comment by bradking** [comment]
I've added a test case in commit 816a57ad63f18b8770291a44ca5b9981037195d7 to cover this.  For C++98 the test is marked as "broken" to record this known limitation.

**Comment by bradking** [comment]
I found that `clang::Sema::MarkFunctionReferenced` skips semantic checking of trivial default constructors.  We simply need to follow up with logic to force such checking.  Fixed by commit 04b836b3114f12bcc849bb126a2310b4a6bf1184.

**Comment by praetorian20** [comment]
Awesome, thank you for looking into it further and finding a fix for C++03 also.
